### ISSUE TITLE
GPU & Core or Thread added

### ISSUE THREAD

**Issue by ashraful1980** [iceland2k14/bsgs#18]
Dear Sir,
Have a nice day and hope you are well.

I had run your software it is amazing speed as 1 peter key per 4 second. It will more speed if you will added GPU & Core or Thread. So i expected to you please execute the option  GPU & Core or Thread; also please added multiple public key as pfile as well as your previous version bsgs_hybrid_gmp.py to the new one bsgs_dll_gmp.py.

Noted you and someone making script include the core or adding thread; reference is as bellow:
https://github.com/iceland2k14/btc_words/blob/main/bip44_words_mcpu.py
https://github.com/Xefrok/BitBruteForce-Wallet/blob/master/seekanddestroy.py

Regards,
Md. Ashraful Alam

**Comment by ashraful1980** [comment]
Dear Sir,
I would like to request you please implement at "bsgs_dll_gmp.py" which detected multiple public key file once a time; it will be helpful for me. At the moment it is count only one; after finished it will going up next...

Regards,
Md. Ashraful Alam

**Comment by ashraful1980** [comment]
Dear Sir,
The bsgs python script is amazing speed.......

![image](https://user-images.githubusercontent.com/52574658/120215123-2e51fb00-c257-11eb-81ba-5afd2fee84c1.png)


Could you please execute option multiple thread. It will find more speed and find puzzle 120 bit private key.

**Comment by iceland2k14** [comment]
Yes mcpu in #TODO list, along with multi pubkey and speed enhancement

**Comment by ashraful1980** [comment]
Sir,
I have 256 gb RAM and speed 10 Quardtrillion meaning 100000 trillion key per second..
I need multiple thread, please execute it...

**Comment by ashraful1980** [comment]
For GPU implementation:

https://www.geeksforgeeks.org/running-python-script-on-gpu/
https://thedatafrog.com/en/articles/boost-python-gpu/
https://numba.pydata.org/

Could you try this please...

**Comment by Manosuper** [comment]
> Dear Sir,
> The bsgs python script is amazing speed.......
> 
> ![image](https://user-images.githubusercontent.com/52574658/120215123-2e51fb00-c257-11eb-81ba-5afd2fee84c1.png)
> 
> Could you please execute option multiple thread. It will find more speed and find puzzle 120 bit private key.

120 puzle. Share 1.2 BTC 50/50 !!! ASAP Please !!!

If someone find a privkey for this pubkey in a range:

1:1124bff742a76f92fa3a76f8f6552f (decimal only 89014422443240977657631574972585263 !!! ) Range is mach less then origginal 120 Puzle range !!!

pubkey:

03c95ba9e6a637e739d13a6d036507ecc37a70c26e68ac7e5f157268311c6c3938
02c95ba9e6a637e739d13a6d036507ecc37a70c26e68ac7e5f157268311c6c3938

search in publey what start with 02 some faster than 03.

you can try any of this pubkeys.

Proof of my succes tests with another ranges and same pubkeys one what start with 03, and anower start with 02:

Working on Pubkey: 049fa3ea651e72c5f0a660c7ae509ad180860477283f539c918a2e43514c29da92237c8751b32e5 90883aa8bd617c60aa6b388aa78381cc7adb10046a33b038598
Using [Number of CPU Threads: 3] [DP size: 10] [MaxStep: 1]
............................................
Scanning Range 0x1 : 0x100000000000000
[55.22 TeraKeys/s][Kang 3072][Count 2^27.92/2^29.06][Elapsed 48s][Dead 0][RAM 25.8MB/44.7MB]
============== KEYFOUND ==============
Kangaroo FOUND PrivateKey : 0xfffffffffffffffffffffffffffffffebaaedce6af48a03bbf59aefd71d64b15

Working on Pubkey: 049fa3ea651e72c5f0a660c7ae509ad180860477283f539c918a2e43514c29da92dc8378ae4cd1a 6f77c557429e839f5594c775587c7e338524effb95bc4fc7697
Using [Number of CPU Threads: 3] [DP size: 10] [MaxStep: 1]
............................................
Scanning Range 0x1 : 0x100000000000000
[42.05 TeraKeys/s][Kang 3072][Count 2^28.68/2^29.06][Elapsed 01:26][Dead 0][RAM 37.7MB/44.7MB]
============== KEYFOUND ==============
Kangaroo FOUND PrivateKey : 0x0000000000000000000000000000000000000000000000000078af8f5e5ff62c
We are spit 1.2 btc 50/50 after with someone who publication in this thread privkey first, all others member sorry... !!!

Lets DO IT ???

Oficila thread - https://bitcointalk.org/index.php?topic=5358408.0

**Comment by ashraful1980** [comment]
Dear Concern,
I could not understand what is your proposal; i understand following proposal that you have kangaroo script along with key range for 120 bit puzzle and once find private key it will share 50/50 it is right. 

I have combo PC "AMD Ryzen 3rd Gen 24 Cores; 48 Threads; 256 GB RAM along with NVIDIA GeForce 3X 3080 GPU" please clear what about your proposal...

I have better performance with JeanLucPons/BSGS, JeanLucPons/Kangaroo and albertobsd/keyhunt...

![image](https://user-images.githubusercontent.com/52574658/132462511-259a1a9d-7b5f-4136-a466-1fba26d789c9.png)

**Comment by Manosuper** [comment]
> Dear Concern,
> I could not understand what is your proposal; i understand following proposal that you have kangaroo script along with key range for 120 bit puzzle and once find private key it will share 50/50 it is right.
> 
> I have combo PC "AMD Ryzen 3rd Gen 24 Cores; 48 Threads; 256 GB RAM along with NVIDIA GeForce 3X 3080 GPU" please clear what about your proposal...
> 
> I have better performance with JeanLucPons/BSGS, JeanLucPons/Kangaroo and albertobsd/keyhunt...
> 
> ![image](https://user-images.githubusercontent.com/52574658/132462511-259a1a9d-7b5f-4136-a466-1fba26d789c9.png)

you will find a private key for one of this this public keys:
search range - 1:1124bff742a76f92fa3a76f8f6552f

03c95ba9e6a637e739d13a6d036507ecc37a70c26e68ac7e5f157268311c6c3938
02c95ba9e6a637e739d13a6d036507ecc37a70c26e68ac7e5f157268311c6c3938



, I will then take the private key that you found, do some manipulations and get a private key. Then I will use it and we will divide 1.2 BTS for two - 50% each, i.e. 0.6 BTC. ????

**Comment by ashraful1980** [comment]
I have check with your pub key; although it was empty no balance yet. I am agreed to your proposal. Please let me know what should i will do....

![image](https://user-images.githubusercontent.com/52574658/132510206-65bd3e1b-1e6c-42d3-9f00-75e31641c6b1.png)
![image](https://user-images.githubusercontent.com/52574658/132510342-26247048-ab50-4509-b79a-9e1c7d68ce95.png)

**Comment by ashraful1980** [comment]
Also please advice search range for public key which you will define 1:1124bff742a76f92fa3a76f8f6552f or 100000000000000000000000000000:1124bff742a76f92fa3a76f8f6552f.

i am waiting for your respond..

**Comment by OSoup** [comment]
I understand that you used the key subtraction and you search on lower key space but if someone will find the private key for the lowered key space he cannot take the btc..only you can because you know the subtraction value. So, who will find the private key have no guarantee that he will receive something.

**Comment by ashraful1980** [comment]
It is not possible to end of reach 1:1124bff742a76f92fa3a76f8f6552f; only possible random which 100000000000000000000000000000:1124bff742a76f92fa3a76f8f6552f.

Thanks ...

**Comment by Manosuper** [comment]
> It is not possible to end of reach 1:1124bff742a76f92fa3a76f8f6552f; only possible random which 100000000000000000000000000000:1124bff742a76f92fa3a76f8f6552f.
> 
> Thanks ...

Ok. Let go to this range, this range can fie too !!!

**Comment by Manosuper** [comment]
> It is not possible to end of reach 1:1124bff742a76f92fa3a76f8f6552f; only possible random which 100000000000000000000000000000:1124bff742a76f92fa3a76f8f6552f.
> 
> Thanks ...

I prepare pubkey for search in range 

1:1124bff742a76f92fa3a76f8f6552f


0313ee32399b1441d0c8cf7caf6d10641b30d280fe1b59fcaa09eb0326cad5cb6f

0213ee32399b1441d0c8cf7caf6d10641b30d280fe1b59fcaa09eb0326cad5cb6f

Regard, I waiting news from you :)

**Comment by ashraful1980** [comment]
The program is running albertobsd/keyhunt..........
Bit range search is 100000000000000000000000000000:1124bff742a76f92fa3a76f8f6552f

**Comment by kpot87** [comment]
And?… what’s news? Have you found ?

**Comment by ashraful1980** [comment]
No.
Till not yet found any result...
